tests/System.IO.Pipelines.Tests/PipeToStreamFacts.cs

```diff
@@ -0,0 +1,168 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.IO.Pipelines.Tests
+{
+    public class PipeToStreamFacts
+    {
+        [Fact]
+        public async Task CopyFromPipeToEnd()
+        {
+            var buffer = new byte[1000];
+            (new Random()).NextBytes(buffer);
+            using (var pipeFactory = new PipeFactory())
+            using (var stream = new MemoryStream())
+            {
+                var pipe = pipeFactory.Create();
+                await pipe.Writer.WriteAsync(buffer);
+                pipe.Writer.Complete();
+                await pipe.Reader.CopyToEndAsync(stream);
+
+                Assert.Throws<InvalidOperationException>(() =>
+                {
+                    var reader = pipe.Reader.ReadAsync();
+                });
+                Assert.Equal(buffer, stream.ToArray());
+            }
+        }
+
+        [Fact] 
+        public async Task CopyFromPipeToEndObserveStreamThrow()
+        {
+            var buffer = new byte[1000];
+            (new Random()).NextBytes(buffer);
+            using (var pipeFactory = new PipeFactory())
+            using (var stream = new FakeExceptionStream())
+            {
+                var pipe = pipeFactory.Create();
+                await pipe.Writer.WriteAsync(buffer);
+                pipe.Writer.Complete();
+                await pipe.Reader.CopyToEndAsync(stream);
+                await Assert.ThrowsAsync<InvalidOperationException>(async () =>
```

Comment:
CopyFromPipeToEndObserveStreamThrow doesn't seem like the appropriate name for what this is testing.

If this was really observing the stream throwing, wouldn't this be an IOException? It seems more like CopyToEndAsync called `IPipeReader.Complete(IOException)`, the the producer was no longer around to observe it. This is a good example of the issue I described in my comment in PipelineConnectionExtensions.cs.

What's really happening here is that `IPipeReader.ReadAsync()` can't be called after  `IPipeReader.Complete()` is with or without an exception.